pkg/controllers/provisioning/batcher.go

```diff
@@ -16,84 +16,77 @@ package provisioning
 
 import (
 	"context"
-	"sync"
 	"time"
 )
 
 var (
 	MaxBatchDuration  = time.Second * 10
 	BatchIdleDuration = time.Second * 1
-	// MaxItemsPerBatch limits the number of items we process at one time to avoid using too much memory
-	MaxItemsPerBatch = 2_000
 )
 
-// Batcher separates a stream of Add(item) calls into windowed slices. The
+// Batcher separates a stream of Trigger() calls into windowed slices. The
 // window is dynamic and will be extended if additional items are added up to a
-// maximum batch duration or maximum items per batch.
+// maximum batch duration.
 type Batcher struct {
-	sync.RWMutex
-	running context.Context
-	queue   chan interface{}
-	gate    context.Context
-	flush   context.CancelFunc
+	running   context.Context
+	trigger   chan struct{}
+	immediate chan struct{}
 }
 
-// NewBatcher is a constructor
+// NewBatcher is a constructor for the Batcher
 func NewBatcher(running context.Context) *Batcher {
-	gate, flush := context.WithCancel(running)
 	return &Batcher{
-		running: running,
-		queue:   make(chan interface{}),
-		gate:    gate,
-		flush:   flush,
+		running:   running,
+		trigger:   make(chan struct{}, 50), // triggering shouldn't block
+		immediate: make(chan struct{}),
 	}
 }
 
-// Add an item to the batch, returning the next gate which the caller may block
-// on. The gate is protected by a read-write mutex, and may be modified by
-// Flush(), which makes a new gate.
-//
-// In rare scenarios, if a goroutine hangs after enqueueing but before acquiring
-// the gate lock, the batch could be flushed, resulting in the pod waiting on
-// the next gate. This will be flushed on the next batch, and may result in
-// delayed retries for the individual pod if the provisioning loop fails. In
-// practice, this won't be encountered because this time window is O(seconds).
-func (b *Batcher) Add(item interface{}) <-chan struct{} {
-	select {
-	case b.queue <- item:
-	case <-b.running.Done():
-	}
-	b.RLock()
-	defer b.RUnlock()
-	return b.gate.Done()
+// Trigger causes the batcher to start a batching window, or extend the current batching window if it hasn't reached the
+// maximum length.
+func (b *Batcher) Trigger() {
+	b.trigger <- struct{}{}
```

Comment:
To remove the need of setting an arbitrary buffered channel value, you could place this in a select with a default case since we won't need to reset the timer if there are triggers already queued.